parsers/US_IPC.py

```diff
@@ -0,0 +1,151 @@
+#!usr/bin/env python3
+
+"""Parser for the Idaho Power Comapny area of the United States."""
+
+from dateutil import parser, tz
+from itertools import groupby
+from logging import getLogger
+import requests
+
+
+# NOTE No pumped storage yet but future ideas can be found at the following url.
+# https://docs.idahopower.com/pdfs/AboutUs/PlanningForFuture/irp/IRP.pdf
+
+# Renewable energy (PURPA) is likely bought with credits from outside the Utility
+# area and not supplied to customers. For that reason those types are commented out. 
+
+PRODUCTION_URL = 'https://api.idahopower.com/Energy/Api/V1/GenerationAndDemand/Subset'
+
+GENERATION_MAPPING = {#'Non-Utility Geothermal': 'geothermal',
+                      'Hydro': 'hydro',
+                      'Coal': 'coal',
+                      'Diesel': 'oil',
+                      #'PURPA/Non-Utility Wind': 'wind',
+                      'Natural Gas': 'gas'
+                      #'PURPA/Non-Utility Solar': 'solar',
+                      #'PURPA Other': 'biomass'
+                      }
+
+TIMEZONE = tz.gettz("America/Boise")
+
+
+def get_data(session=None):
+    """Returns a list of dictionaries."""
+
+    s = session or requests.Session()
+
+    req = requests.get(PRODUCTION_URL)
+    json_data = req.json()
+    raw_data = json_data['list']
+
+    return raw_data
+
+
+def timestamp_converter(timestamp):
+    """Converts a str timestamp into an aware datetime object."""
+
+    dt_naive = parser.parse(timestamp)
+    dt_aware = dt_naive.replace(tzinfo=TIMEZONE)
+
+    return dt_aware
+
+
+def data_processer(raw_data, logger):
+    """
+    Groups dictionaries by datetime key.
+    Removes unneeded keys and logs any new ones.
+    Returns a list of tuples containing (datetime object, dictionary).
+    """
+
+    dt_key = lambda x: x['datetime']
+    grouped = groupby(raw_data, dt_key)
+
+    keys_to_ignore = {'Load', 'Net Purchases', 'Inadvertent','Non-Utility Geothermal',
+                      'PURPA/Non-Utility Wind', 'PURPA/Non-Utility Solar', 'PURPA Other'}
+    known_keys = GENERATION_MAPPING.keys() | keys_to_ignore
+
+    unmapped = set()
+    parsed_data = []
+    for group in grouped:
+        dt = timestamp_converter(group[0])
+        generation = group[1]
+
+        production = {}
+        for gen_type in generation:
+            production[gen_type['name']] = float(gen_type['data'])
```

Comment:
just suggesting another code here:

```python
# None means we should ignore that key
GENERATION_MAPPING = {
  'Hydro': 'hydro',
  'Coal': 'coal',
  'Diesel': 'oil',
  'Natural Gas': 'gas'
  'load': None,
  'Net Purchases': None,
  'Inadvertent': None,
  'Non-Utility Geothermal': None,
  'PURPA/Non-Utility Wind': None,
  'PURPA/Non-Utility Solar': None,
  'PURPA Other': None,
}

[...]
production = {GENERATION_MAPPING[gen_type['name']]: float(gen_type['data'])
              for gen_type in generation
              if GENERATION_MAPPING.get(gen_type['name'])}

unknown_keys = {gen_type['name'] for gen_type in generation} - set(GENERATION_MAPPINGS)
unmapped = unmapped | unknown_keys
[...]
```

it eliminates the need for `current_keys`, `keys_to_remove`, `pop`s, `known_keys` and, I think, makes the code cleaner.

Reply:
How about I wait until Bruno get a response regarding the renewables, then I can update the categories accordingly and use your suggestion.

Reply:
Ok for that: Let start with Idaho power company generation only, and then iterate from here, asking advice if we should mix the **geothermal, wind, solar and other** production to the state-mix. This should become clearer when we investigate neighboring regions. it's a puzzle!